### PR TITLE
REM implementation

### LINKED ISSUE #1447: REM: RFC implementation
Per #815, implement the RFC approved in #1387

**Requirements**:
- Implement `mitigate_measurements`, `execute_with_rem`, `mitigate_executor`, `rem_decorator`.

**Considerations**:
- Circular imports have been challenging with this module because `_typing` uses `MeasurementResult`.

### PR DESCRIPTION
**Pull request by amirebrahimi**
Resolves #1447 

Description
-----------

This is an implementation for the REM RFC approved in #1387. I'm anticipating that some of the validation checks will fail and will fix accordingly.

Checklist
---------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [X] I added unit tests for new code.
- [X] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [X] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [ ] ~I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.~
  This is a separate ticket #1448 
- [ ] ~Added myself / the copyright holder to the AUTHORS file~
  Already added.

License
-------

- [X] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).

### CODE REVIEW

**Comment by natestemen** [comment]
@amirebrahimi it looks like you've bumped numpy which is why mypy is still complaining (we have some work to do to upgrade). Don't feel like you have to do that in this PR, unless you need 1.21.5.

**Comment by amirebrahimi** [comment]
> @amirebrahimi it looks like you've bumped numpy which is why mypy is still complaining (we have some work to do to upgrade). Don't feel like you have to do that in this PR, unless you need 1.21.5.

Thanks @natestemen - I think I might have done it when I was testing with a newer version of Cirq, but I'll revert those changes.

**Comment by amirebrahimi** [comment]
> Hi @amirebrahimi, before reviewing this PR I have a preliminary question. Is there a reason for moving `MeasurementResult` in a new submodule? A new module seems a bit overkilling and it's creation is not directly related to this PR.

Agreed that it was not directly related to this PR. It was the simplest way to resolve circular dependencies.

> A better solution (that we can address in a new PR) is maybe to introduce a new `mitiq.types` module with all Mitiq types including those in `mitiq._typing`. Alternatively, `MeasurementResult` could go into mitiq.executors since it's the output of an executor. Whatever we decide it is probably better to do it in a different PR. Do you agree?

If you would like to go the former, then  think a new PR is good. If you want to go with the latter, then I could try moving it into mitiq.executors in this PR and see if that fixes the issue.

**Comment by andreamari** [comment]
> Agreed that it was not directly related to this PR. It was the simplest way to resolve circular dependencies.
>
Ok, I see. 

> > A better solution (that we can address in a new PR) is maybe to introduce a new `mitiq.types` module with all Mitiq types including those in `mitiq._typing`. Alternatively, `MeasurementResult` could go into mitiq.executors since it's the output of an executor. Whatever we decide it is probably better to do it in a different PR. Do you agree?
> 
> If you would like to go the former, then I think a new PR would be good. If you want to go with the latter, then I could try moving it into mitiq.executors in this PR and see if that fixes the issue.

Any of the following are good solutions (if the fix your circular dependency problem). You can choose what works better for you:

- Just moving `MeasurementResult` into `mitiq._types` (if this is enough, it's fine).
- Moving all Mitiq types into a new `mitiq.types` module.
- Moving `MeasurementResult` into `mitiq.executors`.

In any case, it's better to do this preliminary change in a **different PR** since, if there are will be any problems, they will be easier to isolate. Moreover, after mergin the preliminary PR, this error mitigation PR will become much easier to read (now there are 19 changed files).

**Comment by amirebrahimi** [comment]
> Any of the following are good solutions (if they fix your circular dependency problem). You can choose what works better for you:
> 
> * Just moving `MeasurementResult` into `mitiq._types` (if this is enough, it's fine).
> * Moving all Mitiq types into a new `mitiq.types` module.
> * Moving `MeasurementResult` into `mitiq.executors`.
>
> In any case, it's better to do this preliminary change in a **different PR** since, if there are will be any problems, they will be easier to isolate. Moreover, after mergin the preliminary PR, this error mitigation PR will become much easier to read (now there are 19 changed files).

I tried moving it into `mitiq.executors` first, but this did not work, so I opted to move it into `mitiq._typing` in #1484.

File: .all-contributorsrc [hunk 1]
```diff
@@ -355,6 +355,17 @@
...
+    {
+      "login": "amirebrahimi",
+      "name": "Amir Ebrahimi",
+      "avatar_url": "https://avatars.githubusercontent.com/u/904110?v=4",
+      "profile": "https://github.com/amirebrahimi/",
+      "contributions": [
+        "doc",
+        "test",
+        "code"
+      ]
```
Comment (andreamari) [inline .all-contributorsrc:368]: See this comment: https://github.com/unitaryfund/mitiq/pull/1484#discussion_r968442938

**Comment by amirebrahimi** [comment]
I accidentally merged the other PR, but once that lands to master it should remove that part of the diff.

**Comment by natestemen** [comment]
You can always remove unwanted commits using `git rebase`, or the simpler method of running `git reset --hard HEAD~N` if you want to remove the last _N_ commits.

More info [here](https://stackoverflow.com/questions/1338728/how-do-i-delete-a-commit-from-a-branch).

**Comment by amirebrahimi** [comment]
> You can always remove unwanted commits using `git rebase`, or the simpler method of running `git reset --hard HEAD~N` if you want to remove the last _N_ commits.
> 
> More info [here](https://stackoverflow.com/questions/1338728/how-do-i-delete-a-commit-from-a-branch).

Thanks, I had already pushed and didn't think it was a good thing to re-write history. Perhaps it's not as big a deal as I think though.

File: mitiq/rem/rem.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+) -> MeasurementResult:
+    """Applies the inverse confusion matrix against the noisy measurement
+    result and returns the adjusted measurements.
+
+    Args:
+        noisy_results: The unmitigated ``MeasurementResult``.
+        inverse_confusion_matrix: The inverse confusion matrix to apply to the
+            probability vector estimated with noisy measurement results.
+    """
+    assert noisy_result.qubit_indices is not None
```
Comment (andreamari) [inline mitiq/rem/rem.py:51]: In addition to this assertion, we could raise errors if the size of the inverse confusion matrix is not compatible with the measurement results. E.g. different number of qubits...

File: mitiq/rem/rem.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+from mitiq.executor.executor import Executor
+from mitiq.observable.observable import Observable
+
+MatrixLike = Union[
+    np.ndarray,
+    Iterable[np.ndarray],
+    List[np.ndarray],
+    Sequence[np.ndarray],
+    Tuple[np.ndarray],
+]
```
Comment (andreamari) [inline mitiq/rem/rem.py:36]: Is allowing for this freedom in the type useful/necessary?
If so, we should probably test all types. E.g. does the code work if the matrix is given as a `Tuple[np.ndarray]` ?

Otherwise, the simplest option is to just require that all matrices are `np.ndarray` objects, i.e., without introducing `MatrixLike` at all.

File: mitiq/rem/rem.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+"""Readout Confusion Inversion."""
+
+from typing import Callable, Iterable, List, Optional, Sequence, Tuple, Union
+
+from functools import partial, wraps
+import numpy as np
+
+from cirq.sim import sample_state_vector as sample_prob_dist
+from cirq.qis.states import to_valid_state_vector as to_prob_dist
```
Comment (andreamari) [inline mitiq/rem/rem.py:24]: A state vector and a probability distribution are different mathematical objects. Roughly speaking there is a square root difference between the elements of a state and the element of a probability distribution. 

Isn't this a problem here? It seems that we are using functions that are defined for sampling and generating quantum states for a different task (sampling and generating probability distributions).

**Review by andreamari — COMMENTED** [review]
Hi @amirebrahimi, this is a first step of the review. I looked at the code but I still need to look at the tests.

Meanwhile we can focus on the code which, btw, looks very good and clean.

File: mitiq/rem/rem.py (hunk 2 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py:65]: I've added an additional assert and a new test.

File: mitiq/rem/rem.py (hunk 4 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py]: I've replaced these calls with custom helper methods. You are correct about the semantics of this though. The calls from were Cirq were symmetric, so the result was still the same, but it's unnecessary to turn the measurement result back into a state vector.

**Comment by amirebrahimi** [comment]
Looks like the ubuntu 3.8 test that failed was related to zne; Not sure why that failed though.

File: mitiq/interface/mitiq_cirq/cirq_utils.py [hunk 5]
```diff
@@ -100,3 +104,29 @@ def execute_with_depolarizing_noise(
     rho = simulator.simulate(circuit).final_density_matrix
     expectation = np.real(np.trace(rho @ obs))
     return expectation
+
+
+def generate_inverse_confusion_matrix(
```
Comment (andreamari) [inline mitiq/interface/mitiq_cirq/cirq_utils.py:109]: This function simulates the experimental estimation of the confusion matrix which fine but a quite overkilling.
Given p0 and p1 there should be a fixed analytical solution for the confusion matrix and its inverse without the need to call `cirq.measure_confusion_matrix`.

Fine to keep the function as it is now in this PR, as long as we open a new issue to simplify this function in a future step.

File: mitiq/interface/mitiq_cirq/cirq_utils.py [hunk 6]
```diff
@@ -100,3 +104,29 @@ def execute_with_depolarizing_noise(
...
+    qubits: Sequence["cirq.Qid"],
```
Comment (andreamari) [inline mitiq/interface/mitiq_cirq/cirq_utils.py:110]: For this function the specific choice of qubits looks irrelevant for the returned matrix.
Maybe we can simplify this argument with a simpler `num_qubits: int` ?

File: mitiq/rem/rem.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+from mitiq.executor.executor import Executor
+from mitiq.observable.observable import Observable
+
+
+def to_probability_vector(measurement: np.ndarray) -> np.ndarray:
+    """Convert a raw measurement to a probability vector.
+
+    Args:
+        measurement: A single measurement.
+    """
```
Comment (andreamari) [inline mitiq/rem/rem.py:33]: I think we need a function that returns the probability vector associated to the full list of measurement results.
Something like:

```python
def bitstrings_to_probabiliy_vector(bitstrings: List[Bitstring]) -> np.ndarray:
    """Converts a list of measured bitstrings to a probability vector estimated as
the empirical frequency of each bitstring (ordered with increasing binary value). 
    Args:
        bitstrings: All measured bitstrings.
    """
```

The code of this new function is probably very similar to the code of your function but **averaged** over all bitstrings.

File: mitiq/rem/rem.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    index = int("".join([str(m) for m in measurement]), base=2)
+    pv = np.zeros((2 ** measurement.shape[0]), dtype=np.uint8)
+    pv[index] = 1
+    return pv
+
+
+def sample_probability_vector(probability_vector: np.ndarray) -> np.ndarray:
+    """Sample a probability vector and return a raw measurement.
```
Comment (andreamari) [inline mitiq/rem/rem.py:41]:
```suggestion
    """Sample a probability vector and returns a bitstring.
```

File: mitiq/rem/rem.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    result and returns the adjusted measurements.
+
+    Args:
+        noisy_results: The unmitigated ``MeasurementResult``.
+        inverse_confusion_matrix: The inverse confusion matrix to apply to the
+            probability vector estimated with noisy measurement results.
+    """
+    assert noisy_result.qubit_indices is not None
+    num_qubits = len(noisy_result.qubit_indices)
+    assert inverse_confusion_matrix.shape == (2**num_qubits, 2**num_qubits)
```
Comment (andreamari) [inline mitiq/rem/rem.py:67]: In general we prefer to raise a `ValueError("Some error message")` if the input of a function is not good.

File: mitiq/rem/rem.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+    empirical_prob_dist = np.apply_along_axis(
+        to_probability_vector, 1, noisy_result.asarray
+    )
+
+    adjusted_prob_dist = (inverse_confusion_matrix @ empirical_prob_dist.T).T
+
+    adjusted_result = np.apply_along_axis(
+        sample_probability_vector, 1, adjusted_prob_dist
+    )
```
Comment (andreamari) [inline mitiq/rem/rem.py:77]: Here, you are basically applying REM to each individual measurement.
This is however problematic since, for a single measurement, the adjusted_prob_dist will typically contain negative elements. 
If I am not wrong, the correct way of applying REM is:

-  get a **single** empirical probability distribution associated to all the noisy measurements. 
- Multiply the inverse confusion matrix.
- Hopefully the result doesn't contain negative elements. If it does, we need to think about a way to handle this case.
- Resample new bitstrings from the adjusted distribution. The number of resampled bitstrings should equal to the number of the original noisy bistrings. 
- [Possible alternative to the previous point] Deduce a list of new bitstrings whose frequency best approximate the adjusted distribution.

File: mitiq/rem/rem.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    result = MeasurementResult(adjusted_result, noisy_result.qubit_indices)
+    return result
+
+
+def execute_with_rem(
+    circuit: QPROGRAM,
+    executor: Union[Executor, Callable[[QPROGRAM], MeasurementResult]],
+    inverse_confusion_matrix: np.ndarray,
+    *,
+    observable: Optional[Observable] = None,
```
Comment (andreamari) [inline mitiq/rem/rem.py:88]: For all techniques we always use` circuit, executor, observable` as the first 3 positional arguments. The other keyword-only arguments (after the `*`) are specific to the technique. 
So in this case we need something like:
```python
def execute_with_rem(
    circuit: QPROGRAM,
    executor: Union[Executor, Callable[[QPROGRAM], MeasurementResult]],
    observable: Optional[Observable] = None,    
    *,
    inverse_confusion_matrix: np.ndarray,
```

File: mitiq/rem/rem.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+        probability_vector: A probability vector.
+    """
+    result = np.random.choice(
+        len(probability_vector), size=1, p=probability_vector
+    )[0]
+    value = [int(i) for i in bin(result)[2:]]
+    return np.array(value, dtype=np.uint8)
+
+
+def mitigate_measurements(
```
Comment (andreamari) [inline mitiq/rem/rem.py:53]: It is better if this function and its utilities go into a different file called e.g. `confusion_inversion.py`.
`rem.py` should contain just the high-level functions `execute_with_rem`, `mitigate_executor`, `rem_decorator`.

Note that in Mitiq we use different test files `test_filename.py` associated to different code files. This helps find test of specific functions.

So we probably need a new test file if we create a new file.

**Review by andreamari — COMMENTED** [review]
This is a second round of review about the code. Still not about tests yet.

File: mitiq/interface/mitiq_cirq/cirq_utils.py (hunk 6 above)
Comment (amirebrahimi) [reply mitiq/interface/mitiq_cirq/cirq_utils.py:110]: Unfortunately, as this function currently is it does require the actual Cirq `qubits` to be passed due to line 125:
```
qubit_measures = [[q] for q in qubits]
```

File: mitiq/rem/rem.py (hunk 10 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py:77]: Can you share more about where we get negative elements?

File: mitiq/rem/rem.py (hunk 11 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py:88]: I can make the change, but in this case we should require an inverse confusion matrix or would you rather leave it as a named parameter and throw a value error if one is not passed?

File: mitiq/rem/rem.py (hunk 10 above)
Comment (andreamari) [reply mitiq/rem/rem.py:77]: **The issue**
Even if the confusion matrix contains positive elements, its inverse (inverse confusion matrix) $M^{-1}$ typically contains negative elements. This means that a positive empirical distribution $p$, can be mapped to an adjusted distribution $p' = M^{-1} p$, where some elements of $p'$ are negative. 

Typically negative terms are very small since, by hypothesis, $p$ is not an arbitrary distribution but a noisy distribution having approximately the structure $p=M p_{\rm ideal}$. Unfortunately, negative elements are still possible due to statistical fluctuations or noise miscalibrations.

This effect is much stronger if REM is applied to individual shots (this is the current approach if I am not wrong), instead of applying REM to the averaged empirical distribution (i.e. the vector of frequencies of measured bitstrings).

You can reproduce this _negative probabilities_ issue, if you replace `p0=1` in your tests with something more intermediate e.g.  `p0=0.1`.

**Possible solution**
A possible solution is to clip small negative values of $p'$ to zero and renormalize. The pseudo code I have in mind is:
```python
# Clip negative numbers
adjusted_prob = np.array(adjusted_prob).clip(min=0)
# Renormalize
adjusted_prob = adjusted_prob / np.sum(adjusted_prob)
```

File: mitiq/rem/rem.py (hunk 11 above)
Comment (andreamari) [reply mitiq/rem/rem.py:88]: My proposal is that the argument `inverse_confusion_matrix` should be explicitly passed as a named parameter. Since it doesn't have a default value, the user will get an error if  `inverse_confusion_matrix` is not passed.

We have the same situation with PEC, where `representations` is a necessary parameter but it's not a positional argument.

File: mitiq/interface/mitiq_cirq/cirq_utils.py (hunk 5 above)
Comment (amirebrahimi) [reply mitiq/interface/mitiq_cirq/cirq_utils.py:109]: Added an issue as #1562

File: mitiq/rem/rem.py (hunk 10 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py:77]: Thanks for this explanation. I tried with p0=0.1 and I can see now what you were concerned about. I'll need a little more time with this to correct it.

One question for now:
> get a single empirical probability distribution associated to all the noisy measurements.

I get that this will be a single vector corresponding to the possible outcomes (e.g. measurements for 2 qubits would have four values corresponding to 00, 01, 10, 11). However, should these be raw counts or should they sum to 1?

`[128, 256, 512, 128]` vs. `[0.125, 0.25, 0.5, 0.125]`

File: mitiq/rem/rem.py (hunk 9 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py]: Replaced the asserts with `ValueError`

File: mitiq/rem/rem.py (hunk 11 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py]: Fixed.

File: mitiq/rem/rem.py (hunk 10 above)
Comment (andreamari) [reply mitiq/rem/rem.py]: > Thanks for this explanation. I tried with p0=0.1 and I can see now what you were concerned about. I'll need a little more time with this to correct it.
> 
> One question for now:
> 
> > get a single empirical probability distribution associated to all the noisy measurements.
> 
> I get that this will be a single vector corresponding to the possible outcomes (e.g. measurements for 2 qubits would have four values corresponding to 00, 01, 10, 11). However, should these be raw counts or should they sum to 1?
> 
> `[128, 256, 512, 128]` vs. `[0.125, 0.25, 0.5, 0.125]`

I would maybe normalize such that sum=1. But I don't have a strong preference.  Up to you. 

Note that even if using integer counts (without normalization), after multiplying for $M^{-1}$ the result is a vector with non-integer elements. So I don't see a strong advantage in using integer counts compared to normalized distributions.

File: mitiq/rem/inverse_confusion_matrix.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+def to_probability_vector(measurement: np.ndarray) -> np.ndarray:
+    """Convert a raw measurement to a probability vector.
+
+    Args:
+        measurement: A single measurement.
+    """
+    index = int("".join([str(m) for m in measurement]), base=2)
+    pv = np.zeros((2 ** measurement.shape[0]), dtype=np.uint8)
+    pv[index] = 1
+    return pv
```
Comment (natestemen) [inline mitiq/rem/inverse_confusion_matrix.py:30]: I wonder if this function should go in the `MeasurementResult` class? WDYT?

File: mitiq/rem/rem.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+    result = executor._run([circuit])
+    noisy_result = result[0]
+    assert isinstance(noisy_result, MeasurementResult)
+
+    mitigated_result = mitigate_measurements(
+        noisy_result, inverse_confusion_matrix
+    )
+    assert observable is not None
+    return observable._expectation_from_measurements([mitigated_result])
```
Comment (natestemen) [inline mitiq/rem/rem.py:57]:
```suggestion
    noisy_result = executor._run([circuit])[0]
    assert isinstance(noisy_result, MeasurementResult)

    mitigated_result = mitigate_measurements(
        noisy_result, inverse_confusion_matrix
    )
    assert observable is not None
    return observable._expectation_from_measurements([mitigated_result])
```

We don't have many `assert` statements in our code outside of tests. Is there a reason to use them here over more explicitly raising an error?

**Review by natestemen — COMMENTED** [review]
Nice work Amir. I've left a few small comments.

Also just FYI we may try and integrate with [mthree](https://github.com/Qiskit-Partners/mthree) in the future (https://github.com/unitaryfund/mitiq/issues/1567). Just thought I'd let you know in case you like the measurement error mitigation stuff and want to continue it.

File: mitiq/rem/rem.py (hunk 7 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py]: This has been added.

File: mitiq/rem/rem.py (hunk 10 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py]: Please have a look. I clip the negative values as you suggested.

File: mitiq/rem/inverse_confusion_matrix.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    """Sample a probability vector and returns a bitstring.
+
+    Args:
+        probability_vector: A probability vector.
+    """
+    result = np.random.choice(
+        len(probability_vector), size=1, p=probability_vector
+    )[0]
+    value = [int(i) for i in bin(result)[2:]]
+    return np.array(value, dtype=np.uint8)
```
[the remark this answers is not in the document]
Comment (amirebrahimi) [reply mitiq/rem/inverse_confusion_matrix.py]: This method has been removed.

File: mitiq/rem/inverse_confusion_matrix.py (hunk 13 above)
Comment (amirebrahimi) [reply mitiq/rem/inverse_confusion_matrix.py]: This function was removed.

File: mitiq/rem/rem.py (hunk 14 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py]: I've replaced these with exceptions

**Comment by amirebrahimi** [comment]
> Nice work Amir. I've left a few small comments.
> 
> Also just FYI we may try and integrate with [mthree](https://github.com/Qiskit-Partners/mthree) in the future (#1567). Just thought I'd let you know in case you like the measurement error mitigation stuff and want to continue it.

Thanks, @natestemen. Looks like a useful package. For the time being, I'm mostly busy with my Master's program, so just trying to finish up this feature until I take on any more work.

**Comment by amirebrahimi** [comment]
Any idea why the `asv` check is failing? Seems related to an environment config.
I merged latest and see that pennylane tests are failing.

**Comment by Misty-W** [comment]
> Any idea why the `asv` check is failing? Seems related to an environment config. I merged latest and see that pennylane tests are failing.

Correct, the problem `asv` was in our asv config file. It's fixed now, so once you update or rebase your branch `asv` should pass.

**Comment by natestemen** [comment]
It looks like the tests are failing on python 3.8 due to an error related to using `list` vs `List` in the type hints. Using `list` in type hints is only available in newer versions of python (I thought 3.10, but maybe 3.9 as well). You'll have to use the `List` type from the `typing` module instead.

**Comment by amirebrahimi** [comment]
> It looks like the tests are failing on python 3.8 due to an error related to using `list` vs `List` in the type hints. Using `list` in type hints is only available in newer versions of python (I thought 3.10, but maybe 3.9 as well). You'll have to use the `List` type from the `typing` module instead.

Thanks, Nate. A lesson to keep scrolling further -- I stopped once I saw the errors related to pyquil and pennylane tests and figured it was test instability or something outside of my changes.

So, I'll wait to hear if the current changes are sufficient for the implementation and if so will move on to adding additional tests.

File: mitiq/rem/rem.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    noisy_result = result[0]
+    if not isinstance(noisy_result, MeasurementResult):
+        raise TypeError("Results are not of type MeasurementResult.")
+
+    mitigated_result = mitigate_measurements(
+        noisy_result, inverse_confusion_matrix
+    )
+
+    if observable is None:
+        raise ValueError("An observable is required.")
```
Comment (natestemen) [inline mitiq/rem/rem.py:60]: In the function declaration `observable` defaults to `None` so I would expect this error to be raised very often. I think ideally this function wouldn't require an observable, right?

File: mitiq/rem/rem.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+from mitiq.rem.inverse_confusion_matrix import mitigate_measurements
+
+
+def execute_with_rem(
+    circuit: QPROGRAM,
+    executor: Union[Executor, Callable[[QPROGRAM], MeasurementResult]],
+    observable: Optional[Observable] = None,
+    *,
+    inverse_confusion_matrix: npt.NDArray[np.float64],
+) -> float:
```
Comment (natestemen) [inline mitiq/rem/rem.py:36]: I only recently learning about this `*` syntax to make arguments keyword only! Very cool. Just out of curiosity, is there a reason you want `inverse_confusion_matrix` to be keyword only?

File: mitiq/rem/inverse_confusion_matrix.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    """Converts a list of measured bitstrings to a probability vector estimated
+    as the empirical frequency of each bitstring (ordered with increasing
+    binary value).
+
+    Args:
+        bitstrings: All measured bitstrings.
+    """
+    pv = np.zeros(2 ** len(bitstrings[0]))
+    for bs in bitstrings:
+        index = int("".join([str(bit) for bit in bs]), base=2)
```
Comment (natestemen) [inline mitiq/rem/inverse_confusion_matrix.py:61]: I _think_ this should be marginally faster, and IMO a bit cleaner, but it's also unlikely to make much of an effect since the bitstrings we are likely to deal with are relatively short. Take it or leave it!
```suggestion
        index = int("".join(map(str, bs)), base=2)
```

File: mitiq/rem/inverse_confusion_matrix.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+from typing import List
+import numpy as np
+import numpy.typing as npt
+
+from mitiq._typing import MeasurementResult, Bitstring
+
+
+def sample_probability_vector(
+    probability_vector: npt.NDArray[np.float64], samples: int
+) -> npt.NDArray[np.uint8]:
```
Comment (natestemen) [inline mitiq/rem/inverse_confusion_matrix.py:25]: Should this function returns `Bitstring`s?

**Review by natestemen — COMMENTED** [review]
Nice work Amir. I think this is looking great. I left a few comments, but nothing blocking from adding some more tests if you want to go ahead and do that.

- There are a few docstrings with a missing `Returns:` section to describe what the function returns. Can you make sure all added functions have those, and that they are up to date.
- To check my understanding, the reason we ask the user to pass the `inverse_confusion_matrix` into `execute_with_rem` is because it can be measured once, and then reused many times. If we didn't ask the user to supply it, we would have to calculate it every time `execute_with_rem` is called, which would be very inefficient. Is that right?

**Review by andreamari — COMMENTED** [review]
Hi @amirebrahimi, thanks again for this great contribution! 
The code looks good to me! However I would add the tests I suggested in the comment below. 

If the suggested additional tests don't fail, this is approved from my side!

File: mitiq/rem/rem.py (hunk 17 above)
Comment (andreamari) [reply mitiq/rem/rem.py:36]: @natestemen, sorry to step in, but I asked @amirebrahimi to make `inverse_confusion_matrix`  a keyword-only argument.
The reason is for consistency with all other `execute_with_xxx` functions, e.g. PEC and DDD. The motivation is that we want `observable` to be the third positional argument but with a default value of `None`. 
This fact forces all the arguments on the right side to be keyword-only.

Another reason is that keyword-only arguments are great to avoid position mistakes, especially when there are many options. But this is just my opinion and I am not sure if it is a best practice.

Comment (natestemen) [reply mitiq/rem/rem.py:36]: I'm in favor of it! I was just curious if there was a reason we used it here. Thanks for the info.

File: mitiq/rem/rem.py (hunk 16 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py:60]: Yes that would make sense, but that would be breaking convention from other mitigation techniques (e.g. zne, pec, etc.). I was just following convention. If you are okay with making this required, then I will change it.

File: mitiq/rem/inverse_confusion_matrix.py (hunk 18 above)
Comment (amirebrahimi) [reply mitiq/rem/inverse_confusion_matrix.py:61]: I like that much better. Thanks.

File: mitiq/rem/inverse_confusion_matrix.py (hunk 19 above)
Comment (amirebrahimi) [reply mitiq/rem/inverse_confusion_matrix.py:25]: Yes, that makes more sense since I was calling `.tolist()` where it was used.

**Comment by amirebrahimi** [comment]
> * There are a few docstrings with a missing `Returns:` section to describe what the function returns. Can you make sure all added functions have those, and that they are up to date.

I've fixed these.

> * To check my understanding, the reason we ask the user to pass the `inverse_confusion_matrix` into `execute_with_rem` is because it can be measured once, and then reused many times. If we didn't ask the user to supply it, we would have to calculate it every time `execute_with_rem` is called, which would be very inefficient. Is that right?

Yes, the confusion matrix would be measured every time you wanted to calibrate to the device, then inverted, and re-used many times to mitigate measurements.

File: mitiq/rem/tests/test_rem.py [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+
+    assert second_executor.__doc__ == first_executor.__doc__
+
+
+def test_mitigate_executor():
+    true_rem_value = -2.0
+
+    # test with an executor that completely flips results
+    p0 = 1
+    p1 = 1
```
[the remark this answers is not in the document]
Comment (amirebrahimi) [reply mitiq/rem/tests/test_rem.py]: I've added more tests.

File: mitiq/rem/rem.py (hunk 12 above)
Comment (amirebrahimi) [reply mitiq/rem/rem.py]: I've added a new file: `test_inverse_confusion_matrix.py` with additional unit tests (although we were already covered 100% without them)

**Comment by amirebrahimi** [comment]
Okay, I think I'm ready for a final review @natestemen and @andreamari

File: mitiq/rem/__init__.py [hunk 21]
```diff
@@ -16,3 +16,9 @@
 """Readout error mitigation (REM) techniques."""
 
 from mitiq.rem.post_select import post_select
+from mitiq.rem.inverse_confusion_matrix import mitigate_measurements
+from mitiq.rem.rem import (
+    execute_with_rem,
+    mitigate_executor,
+    rem_decorator,
```
Comment (andreamari) [inline mitiq/rem/__init__.py:23]:
```suggestion
    rem_decorator,
    mitigate_measurements,
```

**Review by andreamari — APPROVED** [review]
LGTM! :champagne:

**Comment by nathanshammah** [comment]
@natestemen does this look good to review / merge? Excited to see this in the codebase.

**Review by natestemen — APPROVED** [review]
Nice work Amir! This is going to be great to have.

**Comment by Misty-W** [comment]
@amirebrahimi when you're ready please click "squash and merge" so it can go out with the upcoming release (milestone 0.21).

Thanks and excited to have this in Mitiq!

**Comment by Misty-W** [comment]
> Thanks, @Misty-W. I don't have write access, so I will need someone on the Mitiq team to merge.

Got it, merged now!
